app/assets/javascripts/shopify_app/storage_access.js

```diff
@@ -0,0 +1,114 @@
+    }
+  });
+  window.parent.postMessage(data, this.redirectInfo.myshopifyUrl);
+}
+
+StorageAccessHelper.prototype.redirectToAppsIndex = function() {
+  window.parent.location.href = this.redirectInfo.myshopifyUrl + '/admin/apps';
+}
+
+StorageAccessHelper.prototype.redirectToAppHome = function() {
+  window.location.href = this.redirectInfo.appHomeUrl;
+}
+
+StorageAccessHelper.prototype.grantedStorageAccess = function() {
+  sessionStorage.setItem('shopify.granted_storage_access', 'true');
+  document.cookie = 'shopify.granted_storage_access=true';
+  this.redirectToAppHome();
+}
+
+StorageAccessHelper.prototype.handleRequestStorageAccess = function() {
+  return document.requestStorageAccess().then(this.grantedStorageAccess.bind(this), this.redirectToAppsIndex.bind(this, ACCESS_DENIED_STATUS));
+}
+
+StorageAccessHelper.prototype.setupRequestStorageAccess = function() {
+  const requestContent = document.querySelector('#RequestStorageAccess');
+  const requestButton = document.querySelector('#TriggerAllowCookiesPrompt');
+
+  requestButton.addEventListener('click', this.handleRequestStorageAccess.bind(this));
+  requestContent.style.display = 'block';
+}
+
+StorageAccessHelper.prototype.handleHasStorageAccess = function() {
+  if (sessionStorage.getItem('shopify.granted_storage_access')) {
+    // If app was classified by ITP and used Storage Access API to acquire access
+    this.redirectToAppHome();
+  } else {
+    // If app has not been classified by ITP and still has storage access
+    this.redirectToAppTLD(ACCESS_GRANTED_STATUS);
+  }
+}
+
+StorageAccessHelper.prototype.handleGetStorageAccess = function() {
+  if (sessionStorage.getItem('shopify.top_level_interaction')) {
+    // If merchant has been redirected to interact with TLD (requirement for prompting request to gain storage access)
+    this.setupRequestStorageAccess();
+  } else {
+    // If merchant has not been redirected to interact with TLD (requirement for prompting request to gain storage access)
+    this.redirectToAppTLD(ACCESS_DENIED_STATUS);
+  }
+}
+
+StorageAccessHelper.prototype.manageStorageAccess = function() {
+  return document.hasStorageAccess().then(function(hasAccess) {
+    if (hasAccess) {
+      this.handleHasStorageAccess();
+    } else {
+      this.handleGetStorageAccess();
+    }
+  }.bind(this));
+}
+
+StorageAccessHelper.prototype.execute = function() {
+  if (ITPHelper.prototype.canPartitionCookies()) {
+    this.setUpCookiePartitioning();
+    return;
+  }
+
+  if (ITPHelper.prototype.userAgentIsAffected()) {
+    this.manageStorageAccess();
+  } else {
+    this.grantedStorageAccess();
+  }
+}
+
+/* ITP 2.0 solution: handles cookie partitioning */
+StorageAccessHelper.prototype.setUpHelper = function() {
+  return new ITPHelper({redirectUrl: window.shopOrigin + "/admin/apps/" + window.apiKey});
+}
+
+StorageAccessHelper.prototype.setCookieAndRedirect = function() {
+  document.cookie = "shopify.cookies_persist=true";
```

Comment:
What's the difference between what happens here and in `partition_cookies.js`? Can it be combined somehow?